python_basics_writing_python_code/python_basics_writing_python_code.md

```diff
@@ -0,0 +1,921 @@
+                "telephone" : "987-654-3210",
+                "email" : "myname@email.com"}
+print(contact_info)
+
+</script>
+</lia-keep>
+</div>
+
+Like lists, dictionaries have some useful methods you can use to access or alter their data:
+
+* `.keys()` returns a list of the dictionary's keys.
+
+    @sage
+    <div class="python_link">
+    <lia-keep>
+    <script type="text/x-sage">
+
+    print(contact_info.keys())
+
+    </script>
+    </lia-keep>
+    </div>
+
+* `.values()` returns a list of the dictionary's values.
+
+    @sage
+    <div class="python_link">
+    <lia-keep>
+    <script type="text/x-sage">
+
+    print(contact_info.values())
+
+    </script>
+    </lia-keep>
+    </div>
+
+* `.items()` returns a list of the dictionary's key-value pairs as tuples (an object that holds multiple values, similar to a list, except that they are unchangeable). This may seem not very useful at first glance, but since different collection objects have different properties, there could be circumstances in which it is more useful to have the data in a different kind of object.
+
+    @sage
+    <div class="python_link">
+    <lia-keep>
+    <script type="text/x-sage">
+
+    print(contact_info.items())
+
+    </script>
+    </lia-keep>
+    </div>
+
+* `.pop()` works similarly for dictionaries as it does for lists, except that it removes an element by its key.
+
+    @sage
+    <div class="python_link">
+    <lia-keep>
+    <script type="text/x-sage">
+
+    print(contact_info.pop("email"))
+print(contact_info)
+
+    </script>
+    </lia-keep>
+    </div>
+
+    <div class="warning">
+    Note that this change is permanent! If you want to keep the original dictionary as-is, you should make a copy of the dictionary by assigning it to a new variable, and then using `.pop()` on the copy. In this example though, all you need to do to restore the dictionary to its original form is re-run the first code cell or refresh the page.
+    </div>
+
+* There isn't a method to add items to a dictionary, but it can be done! You can give the dictionary a new key and assign a value to it (notice that we're using bracket notation again). This is another example of a permanent change to our dictionary.
+
+    @sage
+    <div class="python_link">
+    <lia-keep>
+    <script type="text/x-sage">
+
+    contact_info["work_email"] = "myname@company.com"
+print(contact_info)
+
+    </script>
+    </lia-keep>
+    </div>
+
+  You can also use the same syntax with an exisitng key to assign a new value (note that the new value will **replace** the previous value).
+
+### Quiz: Dictionaries
+
+True or False: Duplicate key-value pairs are allowed in dictionaries.
+
+[( )] True
+[(X)] False
+***
+<div class = "answer">
+While **values** can be repeated in a dictionary, each **key** must be unique. However, you can have multiple values for a key by putting those values in a list.
+</div>
+***
+
+## It's Your Turn!
+
+Ready to practice what you've learned?
```

Comment:
This feels like a pretty big jump for someone who just learned what a list is a few minutes ago. I wonder if there could be some scaffolding of this problem, or maybe a way to ensure that if they go back to read earlier pages they don't lose their work (although we have of course been struggling with the opposite problem with everything hanging around forever...)

Maybe a different phrasing would feel more accurate to the challenge level of the problem, like "can you combine all the things you just learned?" I'm mostly worried about someone doing this lesson, learning a lot, but then getting discouraged because they struggle with the problem and feel like they shouldn't be struggling (even though they should be struggling, that is part of learning!!)

Reply:
Yeah I was worried about this too... I'm leaning toward leaving it out to be honest, but I just love a good "put all the pieces together" practice problem! I'll think on this for the weekend, and either tweak or pull it dependning.

Reply:
+ 1 for a bit of scaffolding